### PR TITLE
Update CUDNN Frontend API to v0.9.1

### PR DESCRIPTION
**Pull request by Tom-Zheng**
### PR types

Function optimization
### PR changes

Others
### Description

- Remove old patches
- Remove workarounds that are no longer needed
- Reduce code redundancy
- Improve cache implementation to avoid multi-thread problems

### CODE REVIEW

**Comment by Tom-Zheng** [comment]
@tianshuo78520a It seems that `PR-CI-Coverage-117-NV` is broken. This PR will need to build with `WITH_CUDNN_FRONTEND=ON`, and tested with `FLAGS_enable_cudnn_frontend=1`. Do you think the new A100 pipeline is ready?

**Comment by Tom-Zheng** [comment]
@tianshuo78520a has confirmed that `PR-CI-OP-benchmark` failure is a random failure and can be manually approved.
@Xreki would you please take a look?

**Comment by Tom-Zheng** [comment]
Adding @jeng1220 @Wong4j for vis

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h [hunk 1]
```diff
@@ -71,58 +79,98 @@ class CudnnFrontendPlanCache {
...
+    // A shared plan being executed by different threads is
+    // generally not safe (for now).
     std::lock_guard<std::mutex> lock(*cache_mutex_);
-    auto engine_config = map_[MakeKey(op_graph, use_addto)];
-    return engine_config;
+    auto &local_map = map_[hasher(std::this_thread::get_id())];
+
+    auto it = local_map.find(GetExtendedFeature(feature, handle));
+    if (it == local_map.end()) {
+      PADDLE_THROW(phi::errors::InvalidArgument(
```
Comment (Xreki) [inline paddle/phi/kernels/autotune/cache_cudnn_frontend.h:94]: 可以直接用`PADDLE_ENFORCE_NE(it, local_map.end(), ...)`

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h [hunk 2]
```diff
@@ -71,58 +79,98 @@ class CudnnFrontendPlanCache {
     return ret;
   }
 
-  cudnn_frontend::ManagedOpaqueDescriptor GetConfig(
-      const cudnn_frontend::OperationGraph& op_graph,
-      cudnnHandle_t handle,
-      bool use_addto = false) {
+  void GetPlan(const cudnn_frontend::feature_vector_t &feature,
```
Comment (Xreki) [inline paddle/phi/kernels/autotune/cache_cudnn_frontend.h:82]: `GetPlan` -> `GetPlanAndWorkspaceSize`

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h [hunk 3]
```diff
@@ -71,58 +79,98 @@ class CudnnFrontendPlanCache {
...
+
+  bool IsStable(const cudnn_frontend::OperationGraph &op_graph,
+                const std::string &tag,
+                cudnnHandle_t handle) {
+    return IsStable(op_graph.getFeatureVector(), tag, handle);
+  }
+
  private:
-  static cudnn_frontend::feature_vector_t MakeKey(
-      const cudnn_frontend::OperationGraph& op_graph, bool use_addto) {
```
Comment (Xreki) [inline paddle/phi/kernels/autotune/cache_cudnn_frontend.h:111]: `use_addto`不需要了吗？

**Comment by Tom-Zheng** [comment]
The pipelines have passed 7 days and need rerun. @tianshuo78520a Could you help rerun `PR-CI-Coverage-117-NV` ?
@Xreki Would you please take some time to review this PR?

**Review by Xreki — APPROVED** [review]
LGTM. 有些小疑问，可以下个PR看下

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h (hunk 2 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/autotune/cache_cudnn_frontend.h:82]: 下个PR修改

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/autotune/cache_cudnn_frontend.h:94]: 下个PR修改

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h (hunk 3 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/autotune/cache_cudnn_frontend.h:111]: CUDNN Frontend 已经将alpha beta加入了feature vector, 故不再需要自己创建feature vector。

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h (hunk 1 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/autotune/cache_cudnn_frontend.h:94]: Fixed at #55026

File: paddle/phi/kernels/autotune/cache_cudnn_frontend.h (hunk 2 above)
Comment (Tom-Zheng) [reply paddle/phi/kernels/autotune/cache_cudnn_frontend.h:82]: Fixed at #55026

**Pull request merged**